### DISCUSSION TITLE
Dagster with SLURM submissions on HPC

### DISCUSSION THREAD

**Discussion by Andrew-S-Rosen** [dagster-io/dagster#14168] [General]
I'll preface this by saying that I'm familiar with workflow tools in general but not yet with dagster.

Most of the tasks I want to run are to be executed on one of a variety of HPC machines, each of which has a SLURM job scheduling system. Can dagster be used to execute tasks in this way? I searched the docs for "SLURM" and didn't get much back except the Dask executor. The Dask executor can indeed be used to submit various job submission scripts, but this is all done locally, right? In other words, one would have to be using dagster on the same machine in which you wish to submit the Dask-generated SLURM job, right? The thing that makes this seem potentially a bit messy is that if you have tasks that you want to run on several different SLURM-hosted machines, you'd have to SSH into each one to execute jobs.

Is there a better/cleaner way to achieve this goal?

**Comment by ghost** [comment]
+1

**Reply by Andrew-S-Rosen** [reply]
I wouldn't hold your breath on this one, but good to see someone else interested. 😄

**Comment by RBrearton** [comment]
I'd also like to add that this would be a very important feature for my use case. We would like to use dagster to orchestrate HPC workflows that are scheduled using slurm.

I appreciate that slurm integration might not be an incredibly common use case in the open source world, but I suspect that there are quite a lot of industry users that would be very interested in this. Given that dagster is VC backed and ultimately needs to make a profit, perhaps it would be reasonable to consider prioritizing a feature like this?

**Comment by ghost** [comment]
I agree.  A lot of people use tools like snakemake and nextflow, but I'm pretty sure if Dagster provided plugins for common HPC schedulers like SLURM, PBS, Torque, etc. there would be a ton of support for that.

**Comment by kazewong** [comment]
+1 on this plugin. While I am not sure how much this is a priority for the dagster team, I am wondering how difficult it would be to add a plugin targeting slurm. `https://github.com/dagster-io/dagster/tree/master/python_modules/libraries/dagster-k8s` is probably an example we can look at, another one is `https://github.com/danielgafni/dagster-ray`.

**Comment by prometheusred** [comment]
I'm doing research for a similar slurm-based, HPC use case now.  And I'm confused by this post.  I see the dask-jobqueue integration can generate sbatch commands... couldn't you just use the appropriate parameterized directives (and appropriate slurm setup) to send jobs to any given part of you slurm cluster?  e.g. to a given partition?  

Anyway, I'll be trying to figure out the way this Dask integration works (e.g. SlurmCluster vs SlurmRunner, also managing logs, managing job status in a dag, etc...) but might also be interested in developing a more custom and direct (non-dask) Slurm integration too as I learn more for my current project....

**Comment by geoHeil** [comment]
I would also be interested in a SLURM integration.

So far I was using dagster pipes with

- local
- databricks
- EMR

for remote compute and bootstrapping of the environment (see https://georgheiler.com/post/paas-as-implementation-detail/ and also the OSS library here https://github.com/ascii-supply-networks/ascii-hydra/tree/main/src/pipelines/ascii_library_demo


I would imagine a solution around:

- dagster pipes (termination + log forwarding + bootstrapping) with the particular HPC environment
- package manager https://pixi.sh/latest/
- packaging of the environement https://github.com/Quantco/pixi-pack

further I would Imagine that dagster would then spin up and start a ray cluster for the assets/computation and then ray would delegate to slurm for the resource management https://docs.ray.io/en/latest/cluster/vms/user-guides/community/slurm.html

**Reply by geoHeil** [reply]
And then ray would implement something like


-  https://docs.dagster.io/guides/build/assets/asset-versioning-and-caching
- https://gafni.dev/projects/sanas-ai-dagster-ray/
- https://dagster.io/blog/unlocking-flexible-pipelines-customizing-asset-decorator
- https://www.youtube.com/watch?v=HPqQSR0BoUQ
- https://www.samsara.com/blog/building-a-modern-machine-learning-platform-with-ray

**Reply by geoHeil** [reply]
What do you think about this approach?

Would this be a good fit?

**Reply by geoHeil** [reply]
Is here anyone around who might be interested in collaborating on an integration of a modern data orchestrator like dagster with supercomputers/slurm? I am thinking of organizing a team to work on that https://www.openhackathons.org/s/siteevent/a0CUP000013Tp8f2AC/se000375 @prometheusred ? @kazewong ? @Andrew-S-Rosen ?

see https://github.com/orgs/ascii-supply-networks/projects/4 for a list of tasks. Even if you do not have time to work hands on-but have experience with SLURM based systems it would be great if you can review the task list if something critical is missing https://github.com/orgs/ascii-supply-networks/projects/4

**Reply by geoHeil** [reply]
And https://github.com/ascii-supply-networks/dagster-slurm to eventually hold the code

**Reply by kazewong** [reply]
I am glad to see someone is taking initiative to make this integration. Right now I am a bit overwhelmed in work so I won't be able to contribute other than reviewing the code from time to time, but I will certainly keep an eye on this.

**Reply by prometheusred** [reply]
I am interested especially if i can make it relevant to my current project.  I’ll have to figure out if / how much time commitment I can actually make based on that.  I haven’t looked closely at your notes or suggested tools (yet) - but I don’t have experience with pixi or ray specifically.  I don’t know much much it matters but my slurm environment is RHEL based and on-prem though I will have integrations with snowflake.

**Reply by geoHeil** [reply]
would it make sense to ahve a quick chat about the envisioned architecture with both of you?

**Reply by prometheusred** [reply]
@geoHeil I'm sorry I've been so slow to respond.  Is it too late to discuss this?

**Reply by geoHeil** [reply]
no totally not - perfect time


quick status update:

we have:

- https://github.com/ascii-supply-networks/dagster-slurm/ basic CI/CD chain
- an idea for the architecture https://github.com/ascii-supply-networks/dagster-slurm/issues/14 @prometheusred please review
- some questions https://github.com/ascii-supply-networks/dagster-slurm/issues/12 the most relevant ones are abit HPC-instance dependent - maybe you @prometheusred can still give some guidance here
- a fully dockerized local mini HPC/slurm cluster for dev
- some documentation pages (still a bit empty but the automation to fill them is set up)

we are working on:

Establishing an initial integration PR is here https://github.com/ascii-supply-networks/dagster-slurm/pull/19

**Reply by geoHeil** [reply]
@prometheusred we are still WIP but something really basic is executable now (limited documentation) https://github.com/ascii-supply-networks/dagster-slurm/pull/19 if you want you can give it a first shot and give feedback


In any case I would be keen to understand how much people would value Session based mode where i.e. not 1 dagster asset -> 1 slurm job but one big slurm job and many assets - possibly with different resources allocated per asset (HET job)

**Reply by geoHeil** [reply]
Update 1: https://github.com/ascii-supply-networks/dagster-slurm/ is now merged in a first version. Please try it and give feedback. Over the next few weeks we intend to add more documentation and perhaps explore https://github.com/orgs/ascii-supply-networks/projects/4?pane=issue&itemId=125777062&issue=ascii-supply-networks%7Cdagster-slurm%7C22

**Reply by geoHeil** [reply]
Update 2: https://github.com/ascii-supply-networks/dagster-slurm/ the documentation is refined now; the first tests on real HPC infrastructure look promising - please explore for yourself and give feedback

**Reply by geoHeil** [reply]
Update 3: VSC5 + Leonardo both are tested successfully now.

**Reply by geoHeil** [reply]
A while ago I shared a link to our project to make HPC systems more accessible - now we have the video from scipy  https://www.youtube.com/watch?v=9uC6tPwMGHE  and are now part of NVIDIAs docs https://docs.rapids.ai/deployment/nightly/examples/rapids-topic-modeling-slurm/notebook/

**Reply by geoHeil** [reply]
FYI: Dagster also now has a `slurm` kind/icon
